### PR TITLE
Update Deposit Contract

### PR DESCRIPTION
**Pull request by nisdas**
This is a part of #2307 
- [x] Update Deposit Contract to spec
- [x] Change how elements are inserted in the deposit trie
- [x] Use deposit hashes now to generate a new trie
- [x] Fix all tests

### CODE REVIEW

File: contracts/deposit-contract/depositContract.go [hunk 1]
```diff
@@ -214,30 +214,30 @@ func (_DepositContract *DepositContractCallerSession) CHAINSTARTFULLDEPOSITTHRES
...
 }
 
-// MAXDEPOSITAMOUNT is a free data retrieval call binding the contract method 0x4c34a982.
+// FULLDEPOSITAMOUNT is a free data retrieval call binding the contract method 0x98faf4ba.
 //
-// Solidity: function MAX_DEPOSIT_AMOUNT() constant returns(uint256 out)
-func (_DepositContract *DepositContractCallerSession) MAXDEPOSITAMOUNT() (*big.Int, error) {
-	return _DepositContract.Contract.MAXDEPOSITAMOUNT(&_DepositContract.CallOpts)
+// Solidity: function FULL_DEPOSIT_AMOUNT() constant returns(uint256 out)
+func (_DepositContract *DepositContractCallerSession) FULLDEPOSITAMOUNT() (*big.Int, error) {
```
Comment (shayzluf) [inline contracts/deposit-contract/depositContract.go]: are we sure we want to have our function names capitalized like that

Comment (nisdas) [reply contracts/deposit-contract/depositContract.go]: @shayzluf this is generated code

File: contracts/deposit-contract/depositContract.go [hunk 2]
```diff
@@ -490,9 +568,9 @@ func (_DepositContract *DepositContractTransactorSession) Drain() (*types.Transa
 	return _DepositContract.Contract.Drain(&_DepositContract.TransactOpts)
 }
 
-// DepositContractChainStartIterator is returned from FilterChainStart and is used to iterate over the raw logs and unpacked data for ChainStart events raised by the DepositContract contract.
-type DepositContractChainStartIterator struct {
-	Event *DepositContractChainStart // Event containing the contract specifics and raw log
+// DepositContractDepositIterator is returned from FilterDeposit and is used to iterate over the raw logs and unpacked data for Deposit events raised by the DepositContract contract.
```
Comment (shayzluf) [inline contracts/deposit-contract/depositContract.go]:
```suggestion
// DepositContractDepositIterator is returned from FilterDeposit and is used to iterate over the raw logs and unpacked data for Deposit events raised by the deposit contract.
```

File: contracts/deposit-contract/depositContract.go [hunk 3]
```diff
@@ -546,42 +624,45 @@ func (it *DepositContractChainStartIterator) Next() bool {
...
 	it.sub.Unsubscribe()
 	return nil
 }
 
-// DepositContractChainStart represents a ChainStart event raised by the DepositContract contract.
-type DepositContractChainStart struct {
-	DepositRoot [32]byte
-	Time        []byte
-	Raw         types.Log // Blockchain specific contextual infos
+// DepositContractDeposit represents a Deposit event raised by the DepositContract contract.
```
Comment (shayzluf) [inline contracts/deposit-contract/depositContract.go]:
```suggestion
// DepositContractDeposit represents a Deposit event raised by the deposit contract.
```

File: contracts/deposit-contract/depositContract.go (hunk 2 above)
Comment (nisdas) [reply contracts/deposit-contract/depositContract.go]: @shayzluf this whole file is generated by abigen

File: contracts/deposit-contract/depositContract.go (hunk 1 above)
Comment (shayzluf) [reply contracts/deposit-contract/depositContract.go]: gotcha

**Pull request merged**